File: cpp/ctbpR-2.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <map>
#include <ctime>
#include <cstdlib>
#include <cstdio>
#include <cstring>
#include <uuid/uuid.h>
#include <boost/filesystem.hpp>
#include <nlohmann/json.hpp>

namespace fs = boost::filesystem;
using json = nlohmann::json;

// Function to create a backup of the file
std::string backup_file(const fs::path& file_path) {
    if (fs::exists(file_path)) {
        std::time_t now = std::time(nullptr);
        char timestamp[20];
        std::strftime(timestamp, sizeof(timestamp), "%Y%m%d_%H%M%S", std::localtime(&now));
        fs::path backup_path = file_path.string() + ".backup_" + timestamp;
        fs::copy_file(file_path, backup_path, fs::copy_option::overwrite_if_exists);
        std::cout << "Backup created at: " << backup_path << std::endl;
        return backup_path.string();
    }
    return "";
}

// Function to determine the storage file location based on the operating system
fs::path get_storage_file() {
    std::string system = 
#ifdef _WIN32
    "Windows";
#elif __APPLE__
    "Darwin";
#elif __linux__
    "Linux";
#else
    "Unknown";
#endif

    if (system == "Windows") {
        char* appdata = std::getenv("APPDATA");
        if (appdata) {
            return fs::path(appdata) / "Cursor" / "User" / "globalStorage" / "storage.json";
        }
    } else if (system == "Darwin") {
        return fs::path(std::getenv("HOME")) / "Library" / "Application Support" / "Cursor" / "User" / "globalStorage" / "storage.json";
    } else if (system == "Linux") {
        return fs::path(std::getenv("HOME")) / ".config" / "Cursor" / "User" / "globalStorage" / "storage.json";
    } else {
        throw std::runtime_error("Unsupported operating system: " + system);
    }
    return fs::path();
}

// Function to generate a random hex string
std::string generate_random_hex(size_t length) {
    std::stringstream ss;
    for (size_t i = 0; i < length; ++i) {
        ss << std::hex << (rand() % 256);
    }
    return ss.str();
}

// Function to generate a UUID
std::string generate_uuid() {
    uuid_t uuid;
    uuid_generate_random(uuid);
    char uuid_str[37];
    uuid_unparse(uuid, uuid_str);
    return std::string(uuid_str);
}

// Function to reset Cursor device IDs
void reset_cursor_id() {
    try {
        fs::path storage_file = get_storage_file();
        fs::create_directories(storage_file.parent_path());

        // Create backup before modifications
        backup_file(storage_file);

        json data;
        if (fs::exists(storage_file)) {
            std::ifstream file(storage_file);
            file >> data;
            std::cout << "Loading configuration from " << storage_file << std::endl;
        } else {
            std::cout << "Storage file not found at " << storage_file << ". Creating new configuration." << std::endl;
        }

        // Generate new random IDs
        std::string machine_id = generate_random_hex(32);
        std::string mac_machine_id = generate_random_hex(32);
        std::string dev_device_id = generate_uuid();

        // Store the original values for logging
        std::map<std::string, std::string> original_ids = {
            {"machineId", data.value("telemetry.machineId", "Not set")},
            {"macMachineId", data.value("telemetry.macMachineId", "Not set")},
            {"devDeviceId", data.value("telemetry.devDeviceId", "Not set")}
        };

        // Update the device IDs
        data["telemetry.machineId"] = machine_id;
        data["telemetry.macMachineId"] = mac_machine_id;
        data["telemetry.devDeviceId"] = dev_device_id;

        // For Cursor v0.47.8, also reset these additional values
        data.erase("cursor.trialStartDate");
        data.erase("cursor.trialReminderShown");
        data.erase("cursor.trialExpired");

        // Write the updated configuration back to the file
        std::ofstream out_file(storage_file);
        out_file << data.dump(2);

        std::cout << "Device IDs successfully reset" << std::endl;

        std::cout << "\n🎉 Cursor v0.47.8 trial successfully reset!\n" << std::endl;
        std::cout << "Original device IDs were:" << std::endl;
        std::cout << json(original_ids).dump(2) << std::endl;
        std::cout << "\nNew device IDs are:" << std::endl;
        std::cout << json({
            {"machineId", machine_id},
            {"macMachineId", mac_machine_id},
            {"devDeviceId", dev_device_id}
        }).dump(2) << std::endl;
        std::cout << "\n✅ Additional trial parameters have been cleared" << std::endl;
        std::cout << "✅ Restart Cursor to apply changes" << std::endl;

    } catch (const std::exception& e) {
        std::cerr << "Error resetting Cursor IDs: " << e.what() << std::endl;
        std::cerr << "\n❌ Error: " << e.what() << std::endl;
        std::cerr << "Please check the script and try again." << std::endl;
    }
}

int main() {
    reset_cursor_id();
    return 0;
}

File: py/ctbpR-2.py
#!/usr/bin/env python3
'''
Cursor Trial Reset Tool
This script resets the device IDs in Cursor's configuration file to generate a new random device ID.
Repository: https://github.com/ultrasev/cursor-reset
Author: @ultrasev
Updated: 23/Mar/2025
Original: 10/Dec/2024
Version: Compatible with Cursor v0.47.8
'''
import json
import os
import shutil
import uuid
import logging
from datetime import datetime
from pathlib import Path
import platform

# Set up logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    datefmt='%Y-%m-%d %H:%M:%S'
)
logger = logging.getLogger(__name__)

def backup_file(file_path: Path):
    """Create a timestamped backup of the given file."""
    if file_path.exists():
        backup_path = f"{file_path}.backup_{datetime.now().strftime('%Y%m%d_%H%M%S')}"
        shutil.copy2(file_path, backup_path)
        logger.info(f"Backup created at: {backup_path}")
        return backup_path
    return None

def get_storage_file():
    """Determine the storage file location based on the operating system."""
    system = platform.system()
    
    if system == "Windows":
        return Path(os.getenv("APPDATA")) / "Cursor" / "User" / "globalStorage" / "storage.json"
    elif system == "Darwin":  # macOS
        return Path(os.path.expanduser("~")) / "Library" / "Application Support" / "Cursor" / "User" / "globalStorage" / "storage.json"
    elif system == "Linux":
        return Path(os.path.expanduser("~")) / ".config" / "Cursor" / "User" / "globalStorage" / "storage.json"
    else:
        raise OSError(f"Unsupported operating system: {system}")

def reset_cursor_id():
    """Reset Cursor device IDs to extend trial period for v0.47.8."""
    try:
        storage_file = get_storage_file()
        storage_file.parent.mkdir(parents=True, exist_ok=True)
        
        # Create backup before modifications
        backup_file(storage_file)
        
        # Load existing data or create new if file doesn't exist
        if not storage_file.exists():
            logger.info(f"Storage file not found at {storage_file}. Creating new configuration.")
            data = {}
        else:
            logger.info(f"Loading configuration from {storage_file}")
            with open(storage_file, 'r', encoding='utf-8') as f:
                data = json.load(f)
        
        # Generate new random IDs
        machine_id = os.urandom(32).hex()
        mac_machine_id = os.urandom(32).hex()
        dev_device_id = str(uuid.uuid4())
        
        # Store the original values for logging
        original_ids = {
            "machineId": data.get("telemetry.machineId", "Not set"),
            "macMachineId": data.get("telemetry.macMachineId", "Not set"),
            "devDeviceId": data.get("telemetry.devDeviceId", "Not set")
        }
        
        # Update the device IDs
        data["telemetry.machineId"] = machine_id
        data["telemetry.macMachineId"] = mac_machine_id
        data["telemetry.devDeviceId"] = dev_device_id
        
        # For Cursor v0.47.8, also reset these additional values
        data.pop("cursor.trialStartDate", None)
        data.pop("cursor.trialReminderShown", None)
        data.pop("cursor.trialExpired", None)
        
        # Write the updated configuration back to the file
        with open(storage_file, 'w', encoding='utf-8') as f:
            json.dump(data, f, indent=2)
        
        logger.info("Device IDs successfully reset")
        
        print("\n🎉 Cursor v0.47.8 trial successfully reset!\n")
        print("Original device IDs were:")
        print(json.dumps(original_ids, indent=2))
        print("\nNew device IDs are:")
        print(json.dumps(
            {
                "machineId": machine_id,
                "macMachineId": mac_machine_id,
                "devDeviceId": dev_device_id,
            },
            indent=2
        ))
        print("\n✅ Additional trial parameters have been cleared")
        print("✅ Restart Cursor to apply changes")
        
    except Exception as e:
        logger.error(f"Error resetting Cursor IDs: {str(e)}")
        print(f"\n❌ Error: {str(e)}")
        print("Please check the script and try again.")

if __name__ == "__main__":
    reset_cursor_id()